server/odc-service/src/main/java/com/oceanbase/odc/service/permission/database/ApplyDatabaseFlowableTask.java

```diff
@@ -0,0 +1,251 @@
+    }
+
+    @Override
+    protected void onTimeout(Long taskId, TaskService taskService) {
+        TaskContextHolder.trace(getTaskCreatorId(taskId, taskService), taskId);
+        log.warn("Apply database permission task timeout, taskId={}", taskId);
+        TaskContextHolder.clear();
+    }
+
+    @Override
+    protected void onProgressUpdate(Long taskId, TaskService taskService) {
+
+    }
+
+    @Override
+    protected boolean cancel(boolean mayInterruptIfRunning, Long taskId, TaskService taskService) {
+        throw new UnsupportedException(ErrorCodes.TaskNotTerminable, null,
+                "The task is not terminable during execution");
+    }
+
+    @Override
+    public boolean isCancelled() {
+        return false;
+    }
+
+    private long getTaskCreatorId(Long taskId, TaskService taskService) {
+        if (this.creatorId != null) {
+            return this.creatorId;
+        }
+        return taskService.detail(taskId).getCreatorId();
+    }
+
+    private void checkResourceAndPermission(ApplyDatabaseParameter parameter) {
+        // Check project still exists
+        Long projectId = parameter.getProject().getId();
+        projectRepository.findById(projectId).orElseThrow(
+                () -> {
+                    log.warn("Project not found, id={}", projectId);
+                    return new NotFoundException(ResourceType.ODC_PROJECT, "id", projectId);
+                });
+        // Check user still exists and is member of the project
+        userRepository.findById(this.creatorId).orElseThrow(
+                () -> {
+                    log.warn("User not found, id={}", this.creatorId);
+                    return new NotFoundException(ResourceType.ODC_USER, "id", this.creatorId);
+                });
+        Set<Long> projectMemberIds = userResourceRoleRepository.findByResourceId(projectId).stream()
+                .map(UserResourceRoleEntity::getUserId).collect(Collectors.toSet());
+        if (!projectMemberIds.contains(this.creatorId)) {
+            log.warn("User not member of project, userId={}, projectId={}", this.creatorId, projectId);
+            throw new IllegalStateException("User not member of project");
+        }
+        // Check databases still exists and belong to the project
+        List<Long> databaseIds =
+                parameter.getDatabases().stream().map(ApplyDatabase::getId).collect(Collectors.toList());
+        Map<Long, DatabaseEntity> id2databaseEntities = databaseRepository.findByIdIn(databaseIds).stream()
+                .collect(Collectors.toMap(DatabaseEntity::getId, d -> d, (d1, d2) -> d1));
+        for (Long id : databaseIds) {
+            if (!id2databaseEntities.containsKey(id)) {
+                log.warn("Database not found, id={}", id);
+                throw new NotFoundException(ResourceType.ODC_DATABASE, "id", id);
+            }
+            if (!id2databaseEntities.get(id).getProjectId().equals(projectId)) {
```

Comment:
It may cause potential NPE as databases could not belong to any project.

Reply:
If it's a white screen operation, the databases that can be selected are the ones that are already associated with the project. However, in order to prevent a packet grabbing attack, it would be more reasonable to add a null value validation.